weasyl/two_factor_auth.py

```diff
@@ -0,0 +1,264 @@
+        WHERE userid = (%(userid)s);
+    """, userid=userid)
+    # Next, generate the recovery codes, up to the defined maximum value
+    tfa_recovery_codes = {security.generate_key(20)}
+    for i in range(0, _TFA_RECOVERY_CODES - 1):
+        tfa_recovery_codes |= {security.generate_key(20)}
+    # Then, insert the codes into the table
+    d.engine.execute("""
+        INSERT INTO twofa_recovery_codes (userid, recovery_code)
+        SELECT (%(userid)s), unnest( (%(tfa_recovery_codes)s) )
+    """, userid=userid, tfa_recovery_codes=list(tfa_recovery_codes))
+    # Finally, return the set of recovery codes to the calling function.
+    return tfa_recovery_codes
+
+
+def is_recovery_code_valid(userid, tfa_code):
+    """
+    Checks the recovery code table for a valid recovery code.
+
+    Determine if a supplied recovery code is present in the recovery code table
+    for a specified userid. If present, consume the code by deleting the record.
+
+    Parameters:
+        userid: The userid of the requesting user.
+        tfa_code: A candidate recovery code to check.
+
+    Returns: Boolean True if the code was valid and has been consumed, Boolean False, otherwise.
+    """
+    # Check to see if the provided code is valid, and consume if so
+    tfa_rc = d.engine.scalar("""
+        DELETE FROM twofa_recovery_codes
+        WHERE userid = (%(userid)s) AND recovery_code = (%(recovery_code)s)
+        RETURNING recovery_code
+    """, userid=userid, recovery_code=tfa_code)
+    # If `tfa_rc` is not None, the code was valid and consumed.
+    if tfa_rc:
+        return True
+    else:
+        return False
+
+
+def is_2fa_enabled(userid):
+    """
+    Check if 2FA is enabled for a specified user.
+
+    Check the ``login.tfa_secret`` field for the tuple identified by ``userid``. If the field is NULL,
+    2FA is not enabled. If it is not null, 2FA is enabled.
+
+    Parameters:
+        userid: The userid to check for 2FA being enabled.
+
+    Returns: Boolean True if 2FA is enabled for ``userid``, otherwise Boolean False.
+    """
+    result = d.engine.scalar("""
+        SELECT twofa_secret
+        FROM login
+        WHERE userid = (%(userid)s)
+    """, userid=userid)
+    if result:
+        return True
+    else:
+        return False
+
+
+def deactivate(userid, tfa_response):
+    """
+    Deactivate 2FA for a specified user.
+
+    Turns off 2FA by nulling-out the ``login.twofa_secret`` field for the user record,
+    and clear any remaining recovery codes.
+
+    Parameters:
+        userid: The user for which 2FA should be disabled.
+        tfa_response: User-supplied response. May be either the Google Authenticator
+        (or other app) supplied code, or a recovery code.
+
+    Returns: Boolean True if 2FA was successfully disabled, otherwise Boolean False if the
+    verification of `tfa_response` failed (bad challenge-response or invalid recovery code).
+    """
+    if verify(userid, tfa_response):
```

Comment:
Above:
``# Sanity checking for length requirement performed in verify() function (6 or 20 length)``